### ISSUE TITLE
Support timer metric tags instead of making a new metric for each method name

### ISSUE THREAD

**Issue by achantavy** [lyft/cartography#1335]
**Feature request template**

*Description*:
> Describe your idea.  Please be detailed.  If a feature request, please
describe the desired behavior, what scenario it enables, and how it
would be used.

Many cartography functions are decorated with `@timeit` (https://github.com/lyft/cartography/blob/e149967cd2750c614d0cb94fced6b1e60870a5cf/cartography/intel/aws/dynamodb.py#L23) so that if a stats server is configured, we can record performance metrics.

We do this by naming the metric the same as the method being called: https://github.com/lyft/cartography/blob/e149967cd2750c614d0cb94fced6b1e60870a5cf/cartography/util.py#L190.

This does not scale because there are many many unique function names, and this approach causes our metrics to be blocked in production or be unnecessarily wasteful.

To avoid this, we should group all of our timer metrics under one single metric name but use tags to differentiate them. The timer metric will be called something like `cartography.timer` and the tag will be the name of the function being called. This way we can observe bottlenecks in performance without being disruptive to our metrics pipelines.

[optional *Relevant Links*:]
> Any extra documentation required to understand the issue.

- https://github.com/lyft/cartography/blob/e149967cd2750c614d0cb94fced6b1e60870a5cf/cartography/util.py#L190
- https://github.com/lyft/cartography/blob/e149967cd2750c614d0cb94fced6b1e60870a5cf/cartography/stats.py#L59
- I'm not up to speed on how to do tagged stats with python but I found this: https://graphite.readthedocs.io/en/latest/tags.html#carbon, maybe someone with more knowledge can chime in.

**Comment by achantavy** [comment]
@ramonpetgrave64 - what you think? Do you know of an oss library in python that lets us do tagged metrics?

**Comment by ramonpetgrave64** [comment]
I'm not aware of any other tagging libs. FYI python stats docs says they [explicitly](https://statsd.readthedocs.io/en/stable/tags.html?highlight=tag) don't support tags.

But we could make a new decorator, based on the the `timeit()` decorator.

```
def make_timer(name: str):
    ...
    # make and return a decorator with this timer
    stats_client = get_stats_client("cartography.timer")
    ...
```

**Comment by achantavy** [comment]
> FYI python stats docs says they [explicitly](https://statsd.readthedocs.io/en/stable/tags.html?highlight=tag) don't support tags.

Yeah.. saw that, so we'll need something different.

> But we could make a new decorator

At the end of the day we still need something that's going to emit the metric with tags though right?

**Comment by ramonpetgrave64** [comment]
It depends on the differentiation needed. What kinds of metric queries were you wanting to write that needed tags?

Instead of tags for the function names, this might work fine.
`cartography.timer.sync_ebs_volumes`

I can't remember how, but it's possible to write a query to sum all of `cartography.timer.*`

**Comment by achantavy** [comment]
I didn't realize the support for tagged metrics with Python was lacking. Blah.

I decided to keep the current approach of multiple metrics and then displaying them all in a chart using the following PromQL:

```
label_replace({__name__=~"production:app:service:.*sync.*:timer:p99"}, "method", "$1", "__name__", "production:app:service:(.*?):timer:p99")'
```

The label_replace() was needed to do regex on the metric name. Then I created a metric alias using `$method` so that the metric itself is labelled with the name of the method being timed.

Thanks for the help!
